libs/parser/src/Parser.php

```diff
@@ -236,9 +253,37 @@ private function parseOrFail(Context $context): iterable
             return $result;
         }
 
-        $token = $context->lastOrdinalToken ?? $context->buffer->current();
+        $token = clone($context->lastOrdinalToken ?? $context->buffer->current());
+
+        if ($this->possibleTokensSearching) {
+            $this->findPossibleTokensForUnexpected($context);
+        }
+
+        throw UnexpectedTokenWithHintsException::fromToken($context->getSource(), $token, null, $this->possibleTokens);
+    }
+
+    /**
+     * @param Context $context
+     * @return void
+     */
+    private function findPossibleTokensForUnexpected(Context $context): void
+    {
+        $problemTokenOffset = $context->lastOrdinalToken->getOffset();
+        $problemTokenKey = 0;
+        while ($context->buffer->get($problemTokenKey)->getOffset() != $problemTokenOffset) {
```

Comment:
There can be any buffer implementation here, not just an ArrayBuffer. In our case, there is an interface where there are no `get`/`set` methods.

Reply:
There are 2 decisions:
1) Add get/set into BufferInterface. Its simple, but it will brake code with manual buffers in exists code.
2) Create BufferGetSetInterface. Check for its implementation in createBuffer method only when possibleTokensSearching option is checked.
I suppose, 2 variant looks better, isnt it?

Reply:
It seems to me that the correct implementation would be to write a decorator over the buffer, like:
```
$buffer = new MutableBuffer(new XxxBufferImplementation());

///

class MutableBuffer implements BufferInterface
{
    private array $overrides = [];

    public function set(int $id, TokenInterface $token): void
    {
        return $this->overrides[$id] = $token;
    }

    public funciton get(int $id): TokenInterface
    {
        return $this->overrides[$id] ?? $this->poll($id);
    }
    
    private funtion poll(int $id): TokenInterface
    {
        $previous = $this->key();
        try {
            $this->seek($id);
            return parent::current();
        } finally {
            $this->seek($previous);
        }
    }
    
    public function current(): TokenInterface
    {
        return $this->overrides[$this->key] ?? parent::current();
    }
}
```

However, this will reduce the speed of the lexer. Though not significantly.

In any case, other implementations are also needed, for example, in the case of processing large data with a non-recursive (with a predetermined nesting depth) grammar, an implementation with [ExtrusiveBuffer](https://github.com/phplrt/phplrt/blob/master/libs/buffer/src/ExtrusiveBuffer.php) makes sense.

Reply:
It is a perfect idea, thank you very much. I added MutableBuffer, it will be used only with possible tokens searching.
Yes, it affects performance. I'm making a DSL for non-programmers, so it has to be user-friendly as possible. Without hints it is sometimes difficult to understand what is wrong.